stream-helpers/README.md

```diff
@@ -0,0 +1,130 @@
+# Stream Helpers
+
+A collection of type-safe stream helpers built on top of
+[Effection](https://github.com/thefrontside/effection) for efficient and
+controlled stream processing.
+
+## Features
+
+### Batch
+
+The `batch` operation groups stream items into batches based on size and/or time
+constraints.
+
+```typescript
+import { batch } from "@effectionx/stream-helpers";
+import { each, run } from "effection";
+
+// Example: Batch by size
+await run(function* () {
+  const stream = batch({ maxSize: 3 })(sourceStream);
+
+  for (const batch of yield* each(stream)) {
+    console.log(batch); // [1, 2, 3], [4, 5, 6], ...
+  }
+});
+
+// Example: Batch by time
+await run(function* () {
+  const stream = batch({ maxTime: 1000 })(sourceStream);
+
+  for (const batch of yield* each(stream)) {
+    console.log(batch); // Items received within 1 second
+    yield* each.next();
+  }
+});
+
+// Example: Combined batching
+await run(function* () {
+  const stream = batch({
+    maxSize: 5,
+    maxTime: 1000,
+  })(sourceStream);
+
+  for (const batch of yield* each(stream)) {
+    console.log(batch); // Up to 5 items within 1 second
+    yield* each.next();
+  }
+});
+```
+
+### Valve
+
+The `valve` operation controls stream flow based on buffer size thresholds.
+
+```typescript
+import { valve } from "@effectionx/stream-helpers";
+import { each, run } from "effection";
+
+await run(function* () {
+  const stream = valve({
+    openAt: 2, // Open valve when buffer size <= 2
+    closeAt: 5, // Close valve when buffer size >= 5
+    open() {
+      console.log("Valve opened");
+    },
+    close() {
+      console.log("Valve closed");
+    },
+  })(sourceStream);
+
+  for (const value of yield* each(stream)) {
+    console.log(value);
+    yield* each.next();
+  }
+});
+```
+
+### Pipe
+
+The `pipe` operation allows composing multiple stream transformations together.
+
+```typescript
+import { valve } from "@effectionx/stream-helpers";
+import { each, run } from "effection";
+// any standard pipe function should work
+import { pipe } from "remeda";
+
+await run(function* () {
+  const valveStream = valve({
+    closeAt: 5,
+    open,
+    close,
+    openAt: 2,
+  });
+
+  // Create a batch processor that batches by size 3
+  const batchStream = batch({ maxSize: 3 });
+
+  // Compose the streams using pipe
+  const composedStream = pipe(
+    sourceStream,
+    valveStream,
+    batchStream,
+  );
+
+  for (const value of yield* each(stream)) {
```

Comment:
I'd put them inline. I think this illustrates why the weird function returning function syntax
```suggestion
  // Compose stream helpers using pipe
  const stream = pipe(
    source,
    valve({ open, close, openAt: 2, closeAt: 5 }),
    batch({ maxSize: 3 }),
  );

  for (const value of yield* each(stream)) {
```